### PR TITLE
Issue 97 Path Length

### PR DESCRIPTION
**Pull request by egolol**
Created the measure and the test.
changed nodal test for measures

### CODE REVIEW

**Comment by egolol** [comment]
1. I think this way is easier, for path length global as I guess is just an average of the path length, we could just inherit PL and take the properties already calculated. You could make an argument that for certain graphs only some part of the calculate is needed, but that can also be coded to this calculate. 

2.  Do you mean: for example  InPL =  mean(Distance, 1);

**Comment by egolol** [comment]
removed the for loop

**Comment by egolol** [comment]
Updated

File: braph/graph/measures/PathLength.m [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+        function m = PathLength(g, varargin)
+            m = m@Measure(g, varargin{:});
+        end
+    end
+    methods (Access = protected)
+        function P =  calculate(m)
+            g = m.getGraph();                          
+            D = g.getMeasure('Distance').getValue();
+            N = g.nodenumber();
+            P = zeros(1, N);
```
Comment (giovannivolpe) [inline braph/graph/measures/PathLength.m]: Change N to n
and P to pl
(in general small letters for vectors and capital for matrices, Du is an exception)

File: braph/graph/measures/PathLength.m [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+            
+            settings = m.getSettings();
+            pathLength_rule = get_from_varargin(0,'PathLengthAvRule', settings{:});
```
Comment (giovannivolpe) [inline braph/graph/measures/PathLength.m]: Add space after 0,

File: braph/graph/measures/PathLength.m [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+            switch lower(pathLength_rule)
+                case {'subgraphs'}                    
+                    for u = 1:1:N
+                        Du = D(:, u);
+                        B(u) = sum(Du(Du~=Inf))/length(Du(Du~=Inf & Du~=0));
```
Comment (giovannivolpe) [inline braph/graph/measures/PathLength.m]: B should be pl

File: braph/graph/measures/PathLength.m (hunk 1 above)
Comment (giovannivolpe) [inline braph/graph/measures/PathLength.m]: pl = zeros(n, 1); 
so already in right format

File: braph/graph/measures/PathLength.m [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+                    end
+                case {'harmonic'}
+                    for u = 1:1:N
+                        Du = D(:,u);
+                        B(u) = harmmean(Du(Du~=Inf & Du~=0));
+                    end
+                otherwise
+                    for u = 1:1:N
+                        Du = D(:,u);
+                        B(u) = mean(Du(Du~=Inf & Du~=0));
```
Comment (giovannivolpe) [inline braph/graph/measures/PathLength.m]: B(u) = mean(Du(Du~=0));

File: braph/graph/measures/PathLength.m [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+            switch lower(pathLength_rule)
+                case {'subgraphs'}                    
+                    for u = 1:1:N
+                        Du = D(:, u);
+                        B(u) = sum(Du(Du~=Inf))/length(Du(Du~=Inf & Du~=0));
+                    end
+                case {'harmonic'}
+                    for u = 1:1:N
+                        Du = D(:,u);
+                        B(u) = harmmean(Du(Du~=Inf & Du~=0));
```
Comment (giovannivolpe) [inline braph/graph/measures/PathLength.m]: B(u) = harmmean(Du(Du~=0));

File: braph/graph/measures/PathLength.m (hunk 3 above)
Comment (giovannivolpe) [inline braph/graph/measures/PathLength.m]: B(u) = mean(Du(Du~=Inf & Du~=0));

**Comment by egolol** [comment]
Updated

File: braph/graph/measures/PathLength.m [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+                        Du = D(:, u);
+                        pl(u) = harmmean(Du(Du~=0));
+                    end
+                otherwise
+                    for u = 1:1:N
+                        Du = D(:, u);
+                        pl(u) = mean(Du(Du~=0));
+                    end
+            end
+            pl(isnan(pl)) = Inf;
```
Comment (giovannivolpe) [inline braph/graph/measures/PathLength.m]: remove this in all files

File: braph/graph/measures/PathLength.m [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+        function measure_class = getClass()
+            measure_class = 'PathLength';
+        end
+        function name = getName()
+            name = 'Path Length';
+        end
+        function description = getDescription()
+            description = [ ...
+                'The path length is the average shortest path lengths of one node to all ' ...
+                'other nodes.' ...
```
Comment (giovannivolpe) [inline braph/graph/measures/PathLength.m]: Make these lines more similar in length

File: braph/graph/measures/test_PathLength.m [hunk 8]
```diff
@@ -0,0 +1,218 @@
...
+p_value = round(p_value,4);
+
+d = Distance(g).getValue();
+value_bct = round(charpath(d), 4);
+
+assert( isequal(p_value(1, 1), value_bct), ...
+    ['BRAPH:PathLength:Subgraphs '], ...
+    ['PathLength is not working for: Subgraphs mean'])
+
+
```
Comment (giovannivolpe) [inline braph/graph/measures/test_PathLength.m]: Add comment that these are modified BCT functions

File: braph/graph/measures/test_PathLength.m [hunk 9]
```diff
@@ -0,0 +1,218 @@
...
+    pathL = g.getMeasure('PathLength');
+    value = pathL.getValue();
+    
+    assert(~isempty(value), ...
+        ['BRAPH:' graph_class ':PathLength'], ...
+        ['PathLength is not calculated for ' graph_class])
+    
+end
+
+%% Test 2: Calculation vs Known Values
```
Comment (giovannivolpe) [inline braph/graph/measures/test_PathLength.m]: make 2 tests withiut using switch

File: braph/graph/measures/test_PathLength.m [hunk 10]
```diff
@@ -0,0 +1,218 @@
+% test PathLength
+graph_class_list = {'GraphBU', 'GraphWU'};
```
Comment (giovannivolpe) [inline braph/graph/measures/test_PathLength.m]: this line goes into test 1

File: braph/graph/measures/InPathLength.m [hunk 11]
```diff
@@ -0,0 +1,75 @@
...
+    methods (Access = protected)
+        function in_path_length =  calculate(m)
+            g = m.getGraph();    
+            
+            settings = g.getSettings();
+            if g.is_measure_calculated('Distance')
+                distance = g.getMeasureValue('Distance');
+            else
+                distance = Distance(g, settings{:}).getValue();
+            end
```
Comment (giovannivolpe) [inline braph/graph/measures/InPathLength.m]: No need for settings = g.getSettings(); anymore, see merging #117 
So, remove:
settings = g.getSettings();
and change to:
distance = Distance(g, g.getSettings()).getValue();

Apply this everywhere

File: braph/graph/measures/OutPathLength.m [hunk 12]
```diff
@@ -0,0 +1,75 @@
...
+    methods (Access = protected)
+        function out_path_length =  calculate(m)
+            g = m.getGraph();
+            
+            settings = g.getSettings();
+            if g.is_measure_calculated('Distance')
+                distance = g.getMeasureValue('Distance');
+            else
+                distance = Distance(g, settings{:}).getValue();
+            end
```
Comment (giovannivolpe) [inline braph/graph/measures/OutPathLength.m]: same

File: braph/graph/measures/OutPathLength.m [hunk 13]
```diff
@@ -0,0 +1,75 @@
...
+            
+            N = g.nodenumber();
+            out_path_length = zeros(N, 1);
+            
+            settings = m.getSettings();
+            pathLength_rule = get_from_varargin(0, 'OutPathLengthAvRule', settings{:});
```
Comment (giovannivolpe) [inline braph/graph/measures/OutPathLength.m]: same

File: braph/graph/measures/PathLength.m [hunk 14]
```diff
@@ -0,0 +1,75 @@
...
+    methods (Access = protected)
+        function path_length =  calculate(m)
+            g = m.getGraph();
+            
+            settings = g.getSettings();
+            if g.is_measure_calculated('Distance')
+                distance = g.getMeasureValue('Distance');
+            else
+                distance = Distance(g, settings{:}).getValue();
+            end
```
Comment (giovannivolpe) [inline braph/graph/measures/PathLength.m]: same

File: braph/graph/measures/PathLength.m [hunk 15]
```diff
@@ -0,0 +1,75 @@
...
+            
+            N = g.nodenumber();
+            path_length = zeros(N, 1);
+            
+            settings = m.getSettings();
+            pathLength_rule = get_from_varargin(0, 'PathLengthAvRule', settings{:});
```
Comment (giovannivolpe) [inline braph/graph/measures/PathLength.m]: same

File: braph/graph/measures/OutPathLength.m [hunk 16]
```diff
@@ -0,0 +1,75 @@
...
+            switch lower(pathLength_rule)
+                case {'subgraphs'}
+                    for u = 1:1:N
+                        Du = distance(u, :);
+                        out_path_length(u) = mean(Du(Du~=0 & Du~=Inf));
+                    end
+                    out_path_length(isnan(out_path_length)) = 0;
```
Comment (giovannivolpe) [inline braph/graph/measures/OutPathLength.m]: Add comment on NaN as in Path Length

File: braph/graph/measures/test_InPathLength.m [hunk 17]
```diff
@@ -0,0 +1,219 @@
...
+                Inf;
+                Inf;
+                Inf;
+                ];
+    end
+    
+    assert( isequal(p_value, known_solution), ...
+        ['BRAPH:InPathLength: ' graph_class], ...
+        ['InPathLength is not working for: ' graph_class ])
+end
```
Comment (giovannivolpe) [inline braph/graph/measures/test_InPathLength.m]: Split into two tests.
Add also extra matrix where the result is not Inf

File: braph/graph/measures/test_InPathLength.m [hunk 18]
```diff
@@ -0,0 +1,219 @@
...
+p_value = round(p_value,4);
+
+d = Distance(g).getValue();
+value_bct = round(charpath(d), 4);
+
+assert( isequal(p_value(1, 1), value_bct), ...
+    ['BRAPH:InPathLength:Subgraphs '], ...
+    ['InPathLength is not working for: BCT'])
+
+
```
Comment (giovannivolpe) [inline braph/graph/measures/test_InPathLength.m]: Add comment that functions from BCT

File: braph/graph/measures/test_OutPathLength.m [hunk 19]
```diff
@@ -0,0 +1,212 @@
+% test OutPathLength
```
Comment (giovannivolpe) [inline braph/graph/measures/test_OutPathLength.m]: same comments as in test_InPathLength

File: braph/graph/measures/test_PathLength.m [hunk 20]
```diff
@@ -0,0 +1,214 @@
+% test PathLength
```
Comment (giovannivolpe) [inline braph/graph/measures/test_PathLength.m]: same comments as in test_InPathLength

File: braph/graph/measures/test_PathLength.m (hunk 10 above)
Comment (giovannivolpe) [reply braph/graph/measures/test_PathLength.m]: Apply this to all, as you don't need it in the other tests.

**Comment by giovannivolpe** [comment]
It's good, only very small things remaining.

File: braph/graph/measures/test_InPathLength.m (hunk 17 above)
Comment (egolol) [reply braph/graph/measures/test_InPathLength.m]: I dont agree with separating, its code duplication

Comment (giovannivolpe) [reply braph/graph/measures/test_InPathLength.m]: Ok, leave it as it is now

File: braph/graph/measures/test_InPathLength.m (hunk 18 above)
Comment (egolol) [reply braph/graph/measures/test_InPathLength.m]: I dont understand the review

Comment (giovannivolpe) [reply braph/graph/measures/test_InPathLength.m]: Small thing, just add a line before the BCT functions saying:
% Functions from 2019_03_03_BCT
or 
% Functions adapted from 2019_03_03_BCT
depending on the case

**Comment by egolol** [comment]
well the problem is in get_from_varargin you are comparing wrong strcmpi(varargin{n}, handle) 
image

**Comment by egolol** [comment]
I was working in an old version

**Pull request merged**